src/theory/booleans/proof_checker.cpp

```diff
@@ -159,23 +187,50 @@ Node BoolProofRuleChecker::checkInternal(PfRule id,
   if (id == PfRule::CHAIN_RESOLUTION)
   {
     Assert(children.size() > 1);
-    Assert(args.size() == children.size() - 1);
+    Assert(args.size() == 2 * (children.size() - 1));
     Trace("bool-pfcheck") << "chain_res:\n" << push;
+    NodeManager* nm = NodeManager::currentNM();
+    Node trueNode = nm->mkConst(true);
+    Node falseNode = nm->mkConst(false);
     std::vector<Node> clauseNodes;
     for (unsigned i = 0, childrenSize = children.size(); i < childrenSize; ++i)
     {
-      std::unordered_set<Node, NodeHashFunction> elim;
+      std::unordered_map<Node, unsigned, NodeHashFunction> elim;
       // literals to be removed from "first" clause
       if (i < childrenSize - 1)
       {
-        elim.insert(args.begin() + i, args.end());
+        for (unsigned j = (2*i), argsSize = args.size(); j < argsSize; j = j + 2)
+        {
+          // whether pivot should occur as is or negated depends on the id of
```

Comment:
Maybe it's too early in the morning, but I can't follow what you're doing here.

In `elim`, you are collecting all pivot nodes (negated or not) starting at the current `i`, and then you add the current pivot node (negated). Finally, you collect all literals of the current clause and add them (minus what is in `elim`) into `clauseNodes`.

So what confuses me is the following: why do you compute `elim` in every loop from scratch? It seems to me that this removes pivots multiple times from the clauses. Assume you have three pivot (with four clauses) you would put the last pivot three times in `elim`. Why should the pivot used for the resolution of the third and fourth clause be eliminated from the first clause?

Reply:
You're right. This is yet another misconception because I inadvertently wrote a checker for resolution + some factoring when I should have written one for just resolution. This is good though because the checking will be much simpler now. Thanks for noticing!

Reply:
After internal discussion we've decided to change the chain resolution rule in another PR, as significant changes will be necessary.